### PR TITLE
KEYCLOAK-953: add allowing user to delete his own account feature

### PR DESCRIPTION
**Pull request by zak905**

### CODE REVIEW

**Comment by KonstantinCodes** [comment]
This feature is very much appreciated. It will help us tremendously. Thank You!

**Comment by zak905** [comment]
anybody can help with the `crossdc-server` tests, they seem to be the only ones failing. I am not sure how I can debug, since the tests seem to spin up lots of resources..infinispan servers, database, load balancer,...etc

**Comment by zak905** [comment]
I saw the same error on this PR https://github.com/keycloak/keycloak/pull/6237#partial-pull-merging, which means that the tests failures may not be related to the changes I introduced. Waiting for somebody to confirm...

**Comment by zak905** [comment]
is there anyway to skip `crossdc-server` tests ?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/AbstractKeycloakTest.java [hunk 1]
```diff
@@ -34,9 +34,11 @@
 import org.keycloak.admin.client.resource.UsersResource;
 import org.keycloak.common.util.KeycloakUriBuilder;
 import org.keycloak.common.util.Time;
+import org.keycloak.models.AdminRoles;
```
Comment (zak905) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/AbstractKeycloakTest.java]: Will remove these imports, I was trying to fix the crossdc-server tests, I left them there

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/crossdc/SessionExpirationCrossDCTest.java [hunk 2]
```diff
@@ -72,6 +73,18 @@
     private int authSessions02;
 
 
+    @Override
```
Comment (zak905) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/crossdc/SessionExpirationCrossDCTest.java]: TO be removed: It was an attempt to fix the crossdc-server tests without success

**Comment by stianst** [comment]
You can ignore the failing crossdc-test as they are unstable, so most likely not your fault.

**Comment by zak905** [comment]
thanks for the confirmation, I thought so

File: model/jpa/src/main/java/org/keycloak/models/jpa/entities/RealmEntity.java [hunk 3]
```diff
@@ -80,6 +80,8 @@
     protected boolean duplicateEmailsAllowed;
     @Column(name="REMEMBER_ME")
     protected boolean rememberMe;
+    @Column(name="USER_DELETE_OWN_ACCOUNT_ALLOWED")
```
Comment (stianst) [inline model/jpa/src/main/java/org/keycloak/models/jpa/entities/RealmEntity.java]: We used to have columns for every setting, but it doesn't scale very well and requires a lot of updates to DB schema. Hence we introduce the generic realm attributes. Please change this to use realm attributes. In the JSON representation it should be a field, but in the realm entity it should be a realm attribute.

File: model/jpa/src/main/resources/META-INF/jpa-changelog-8.0.0.xml [hunk 4]
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Comment (stianst) [inline model/jpa/src/main/resources/META-INF/jpa-changelog-8.0.0.xml]: See above - DB schema changes is not neeed for this feature

**Review by stianst — CHANGES_REQUESTED** [review]
Deleting the user account should rather be driven by an application initiated action as that will make it possible to require the user to re-authenticate. The action should also ask the user to confirm deleting the account prior to actually deleting. Using an AIA will also make it easy to add this capability to the new account console.

So the flow should be:

* Click [delete] in account console
* User is redirected to login page with kc_action=delete_user
* User is required to re-authenticate (we are working on adding this to AIAs currently)
* Delete user action shows a form asking the user to confirm deleting the account. It should have a cancel button like other AIAs.

I'm not convinced it should be on a separate page, but perhaps rather just added to the current profile page. Would like input from others on this.

**Comment by zak905** [comment]
Hi @stianst, 

Thanks for the review. It makes sens that the user should be re-prompted to authenticate when deleting his own account, since this is a sensitive operation. 

I saw that AIAs has only been added very recently to keycloak, and that no operation in the user account page integrates them yet (e.g update password). I saw also that https://issues.jboss.org/browse/KEYCLOAK-11549 is tightly related to what this ticket need, therefore I would not want to reinvent the wheel or add something that will conflict with it. until KEYCLOAK-11549 is completed, I suggest that we merge this one, and create another ticket (with subtasks maybe ) that will add re-authentication through AIA to delete account, and also to other functionalities of the account page (like updating password or profile, and adding otp) at the same time, I will be happy to contribute again. It would not make sens to add it to only delete account, and not the others which would be out of the scope of this ticket anyway. I think the changes in the PR fullfills KEYCLOAK-953 initial requirements, which is just the initial raw functionality of deleting user's account.

**Comment by zak905** [comment]
I included the changes requested above (changing from a DB colum to a realm attribute). Here is some testing scenarios:
- running org.keycloak.testsuite.KeycloakServer on an IDE or using mvn
- creating a realm
- activating the "Account deletion by user" from Realm settings -> general tab
- adding the "delete-own-account" to realm default roles by going to Roles -> Default Roles 
- creating a user and clicking on impersonate user 
- going Delete Account, and clicking on delete account 
- click on delete account, you will be logged out immediately
- go back to the admin, and find that the user has been deleted
- deactivate the "Account deletion by user", and create a new user, and go to impersonate again, you will find that the Delete Account menu section is not visible. 
- repeat again without the realm "delete-own-account" role, you will receive a error message on the "Delete Account" page that says "not enough privileges, contact admin"

**Comment by stianst** [comment]
AIAs are designed for the new account console. Converting this to use an AIA will make it simpler to address the two requirements:

* Ask user if they are sure they want to delete the account first
* Ask user to re-authenticate (AIA doesn't support this right now, but will do once KEYCLOAK-11549 is added)

The alternative to using AIAs is to implement the above checks in the old account console, just to convert them to AIA later, which will then also not be usable to the new account console since it relies on AIAs for actions like these.

File: server-spi-private/src/main/java/org/keycloak/models/AccountRoles.java [hunk 5]
```diff
@@ -29,6 +29,7 @@
     String VIEW_APPLICATIONS = "view-applications";
     String VIEW_CONSENT = "view-consent";
     String MANAGE_CONSENT = "manage-consent";
+    String DELETE_ACCOUNT = "delete-own-account";
```
Comment (stianst) [inline server-spi-private/src/main/java/org/keycloak/models/AccountRoles.java]: Should just be "delete-account" to make it consistent with other roles. It is "manage-account", not "manage-own-account"

File: services/src/main/java/org/keycloak/forms/account/freemarker/model/RealmBean.java [hunk 6]
```diff
@@ -72,4 +72,8 @@ public boolean isRegistrationEmailAsUsername() {
     public boolean isUserManagedAccessAllowed() {
         return realm.isUserManagedAccessAllowed();
     }
+
+    public boolean isUserDeleteOwnAccountAllowed() {
```
Comment (stianst) [inline services/src/main/java/org/keycloak/forms/account/freemarker/model/RealmBean.java]: This option is not needed at all. By default users shouldn't have the delete-account role associated with them and if they have it then they can delete their account, so there is no need for an additional config option in the realm.

File: services/src/main/java/org/keycloak/services/resources/account/AccountFormService.java [hunk 7]
```diff
@@ -1090,5 +1141,12 @@ private void csrfCheck(final MultivaluedMap<String, String> formData) {
         }
     }
 
+    private void checkRealmAndClientRoles() {
```
Comment (stianst) [inline services/src/main/java/org/keycloak/services/resources/account/AccountFormService.java]: Why is this a separate method? Why is it checking both client and realm roles? There shouldn't be a realm level delete account role, only a client level associated with account client.

File: services/src/test/java/org/keycloak/vault/PlainTextVaultProviderFactoryTest.java [hunk 8]
```diff
@@ -243,6 +243,16 @@ public void setUserManagedAccessAllowed(boolean userManagedAccessAllowed) {
           
         }
 
+        @Override
```
Comment (stianst) [inline services/src/test/java/org/keycloak/vault/PlainTextVaultProviderFactoryTest.java]: Not sure what's going on here.

**Review by stianst — CHANGES_REQUESTED** [review]
Remaining work:
* Convert to AIA
* Start a discussion on dev ML if there should be a separate delete account page, or if it should be on the profile page
* Remove realm option as it's not needed since delete account role has to be associated with user before the user can delete the account
* Rename role to delete-account
* delete-account should not be a realm role, seems PR supports it as both realm role and client role
* Missing migration code adding role to existing realms in the DB

**Pull request closed without merging**